presto-orc/src/main/java/com/facebook/presto/orc/reader/TimestampSelectiveStreamReader.java

```diff
@@ -0,0 +1,430 @@
+    }
+
+    @Override
+    public Block getBlock(int[] positions, int positionCount)
+    {
+        checkArgument(outputPositionCount > 0, "outputPositionCount must be greater than zero");
+        checkState(outputRequired, "This stream reader doesn't produce output");
+        checkState(positionCount <= outputPositionCount, "Not enough values");
+        checkState(!valuesInUse, "BlockLease hasn't been closed yet");
+
+        if (allNulls) {
+            return new RunLengthEncodedBlock(NULL_BLOCK, outputPositionCount);
+        }
+
+        boolean includeNulls = nullsAllowed && presentStream != null;
+        if (positionCount == outputPositionCount) {
+            Block block = new LongArrayBlock(positionCount, Optional.ofNullable(includeNulls ? nulls : null), values);
+            nulls = null;
+            values = null;
+            return block;
+        }
+
+        long[] valuesCopy = new long[positionCount];
+        boolean[] nullsCopy = null;
+        if (includeNulls) {
+            nullsCopy = new boolean[positionCount];
+        }
+
+        int positionIndex = 0;
+        int nextPosition = positions[positionIndex];
+        for (int i = 0; i < outputPositionCount; i++) {
+            if (outputPositions[i] < nextPosition) {
+                continue;
+            }
+
+            assert outputPositions[i] == nextPosition;
+
+            valuesCopy[positionIndex] = this.values[i];
+            if (nullsCopy != null) {
+                nullsCopy[positionIndex] = this.nulls[i];
+            }
+
+            positionIndex++;
+            if (positionIndex >= positionCount) {
+                break;
+            }
+
+            nextPosition = positions[positionIndex];
+        }
+
+        return new LongArrayBlock(positionCount, Optional.ofNullable(nullsCopy), valuesCopy);
+    }
+
+    @Override
+    public BlockLease getBlockView(int[] positions, int positionCount)
+    {
+        checkArgument(outputPositionCount > 0, "outputPositionCount must be greater than zero");
+        checkState(outputRequired, "This stream reader doesn't produce output");
+        checkState(positionCount <= outputPositionCount, "Not enough values");
+        checkState(!valuesInUse, "BlockLease hasn't been closed yet");
+
+        if (allNulls) {
+            return newLease(new RunLengthEncodedBlock(NULL_BLOCK, outputPositionCount));
+        }
+
+        boolean includeNulls = nullsAllowed && presentStream != null;
+        if (positionCount != outputPositionCount) {
+            compactValues(positions, positionCount, includeNulls);
+        }
+
+        return newLease(new LongArrayBlock(positionCount, Optional.ofNullable(includeNulls ? nulls : null), values));
+    }
+
+    private BlockLease newLease(Block block)
+    {
+        valuesInUse = true;
+        return ClosingBlockLease.newLease(block, () -> valuesInUse = false);
+    }
+
+    private void compactValues(int[] positions, int positionCount, boolean compactNulls)
```

Comment:
There are multiple readers that store data in array of longs. Would it be make sense to extract this method and logic in getBlock into a utility class to avoid copy-paste?

Reply:
would like to take care of this in a new commit, once the scalars are in I can do a quick refactor there. what do you think?